src/source/source_cache.js

```diff
@@ -168,7 +168,7 @@ class SourceCache extends Evented {
             this._source.prepare();
         }
 
-        this._state.coalesceChanges(this._tiles);
+        this._state.coalesceChanges(this._tiles, this.map ? this.map.painter : null);
```

Comment:
@anandthakker, in [#6236(comment)](https://github.com/mapbox/mapbox-gl-js/pull/6263#discussion_r177534083), we discussed (and agreed upon) : 
>-  move the responsibility for calling Tile#updateFeatureState during prepare() directly into SourceFeatureState#coalesceChanges() (i.e., take a set of tiles as a parameter to coalesceChanges)?
> - Add a SourceFeatureState#initializeTileState(Tile) method and replace the other two cases of tile.updateFeatureState(this._state.state); with this._state.initializeTileState(tile)?

What do you think of moving back the tile-specific code from `this._state.coalesceChanges` into `SourceCache`, to reduce the `painter` spaghetti created here. 
Not a burning issue though.

Reply:
I'm thinking we should keep this as-is, because I expect this bit of  🍝 to be temporary: via https://github.com/mapbox/mapbox-gl-js/issues/4875 and https://github.com/mapbox/mapbox-gl-js/issues/4012, I'm hoping we'll end up in a place where each tile can easily hold on to a view of the style layers that were used to lay it out, and that will let us eliminate the need to thread `painter` through here.